core/templates/domain/statistics/LearnerActionObjectFactory.ts

```diff
@@ -20,96 +20,94 @@
-    return {
-      action_type: this.actionType,
-      action_customization_args: this.actionCustomizationArgs,
-      schema_version: this.schemaVersion,
-    };
-  }
+export interface ILearnerActionExplorationStart {
+  // NOTE TO DEVELOPERS: The following two definitions are NOT typos! These are
+  // discriminated unions, a feature of TypeScript. Code will fail to compile if
+  // the compiler can not deduce that action_type and schema_version are set to
+  // the following values. For details see:
+  // https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions eslint-disable max-len
+  'action_type': 'ExplorationStart';
+  'schema_version': 1;
+  'action_customization_args': IExplorationStartCustomizationArgs;
 }
 
+export interface ILearnerActionAnswerSubmit {
+  // NOTE TO DEVELOPERS: The following two definitions are NOT typos! These are
+  // discriminated unions, a feature of TypeScript. Code will fail to compile if
+  // the compiler can not deduce that action_type and schema_version are set to
+  // the following values. For details see:
+  // https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions eslint-disable max-len
+  'action_type': 'AnswerSubmit';
+  'schema_version': 1;
+  'action_customization_args': IAnswerSubmitCustomizationArgs;
+}
+
+export interface ILearnerActionExplorationQuit {
+  // NOTE TO DEVELOPERS: The following two definitions are NOT typos! These are
+  // discriminated unions, a feature of TypeScript. Code will fail to compile if
+  // the compiler can not deduce that action_type and schema_version are set to
+  // the following values. For details see:
+  // https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions eslint-disable max-len
+  'action_type': 'ExplorationQuit';
+  'schema_version': 1;
+  'action_customization_args': IExplorationQuitCustomizationArgs;
+}
+
+export type ILearnerAction = (
+  ILearnerActionExplorationStart |
+  ILearnerActionAnswerSubmit |
+  ILearnerActionExplorationQuit);
+
 @Injectable({
   providedIn: 'root'
 })
 export class LearnerActionObjectFactory {
-  /**
-   * @property {string} actionType - type of an action
-   * @property {Object.<string, *>} actionCustomizationArgs - customization
-   *   dict for an action
-   * @property {number} [schemaVersion=LEARNER_ACTION_SCHEMA_LATEST_VERSION]
-   *   - schema version of the class instance.
-   * @returns {LearnerAction}
-   */
-  createNew(
-      actionType: string, actionCustomizationArgs: any,
-      schemaVersion: number): LearnerAction {
-    schemaVersion = schemaVersion ||
-      StatisticsDomainConstants.LEARNER_ACTION_SCHEMA_LATEST_VERSION;
-    return new LearnerAction(
-      actionType, actionCustomizationArgs, schemaVersion);
+  createExplorationStartAction(
+      actionCustomizationArgs:
+        IExplorationStartCustomizationArgs): ILearnerActionExplorationStart {
+    return {
+      action_type: 'ExplorationStart',
```

Comment:
Why don't you make actual domain object for these rather than using dicts? I actually prefer it the way it was before.

Or is the point that these dicts are not mutable? If so, then this may be a feasible alternative, but I would like us to have a single convention for "domain objects" that applies across the entire codebase, rather than doing something piecemeal here. /cc @nishantwrp @vojtechjelinek for advice.